### PR TITLE
[PAN-2499] debug trace transaction

### PR DESCRIPTION
**Pull request by abdelhamidbakhta**
## PR description
The `REVERT` provides the ability to return a reason string .
We should get this reason when calling `debug_traceTransaction`
## Fixed Issue(s)

fixes #PAN-2499

### CODE REVIEW

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/operations/RevertOperation.java [hunk 1]
```diff
@@ -36,8 +40,10 @@ public Gas cost(final MessageFrame frame) {
   public void execute(final MessageFrame frame) {
     final UInt256 from = frame.popStackItem().asUInt256();
     final UInt256 length = frame.popStackItem().asUInt256();
-
-    frame.setOutputData(frame.readMemory(from, length));
+    BytesValue reason = frame.readMemory(from, length);
+    frame.setOutputData(reason);
+    String reasonMessage = new String(reason.extractArray(), CHARSET);
```
Comment (shemnon) [inline ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/operations/RevertOperation.java]: Strings in solidity are UTF-8 (https://solidity.readthedocs.io/en/v0.5.3/types.html#dynamically-sized-byte-array) which is where these values will come from.  Use `java.nio.charset.StandardCharsets.UTF_8`.

File: pantheon/src/main/java/tech/pegasys/pantheon/cli/PantheonCommand.java [hunk 2]
```diff
@@ -129,29 +125,6 @@
 
   private CommandLine commandLine;
 
-  public static class RpcApisConverter implements ITypeConverter<RpcApi> {
```
Comment (shemnon) [inline pantheon/src/main/java/tech/pegasys/pantheon/cli/PantheonCommand.java]: Moving these classes out is orthagonal to this PR. Please do this in another PR.

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/debug/TraceFrame.java [hunk 3]
```diff
@@ -54,6 +78,7 @@ public TraceFrame(
     this.stack = stack;
     this.memory = memory;
     this.storage = storage;
+    this.reason = null;
```
[the remark this answers is not in the document]
Comment (abdelhamidbakhta) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/debug/TraceFrame.java]: Done.

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/operations/RevertOperation.java (hunk 1 above)
Comment (abdelhamidbakhta) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/operations/RevertOperation.java]: Switched to UTF-8

File: pantheon/src/main/java/tech/pegasys/pantheon/cli/PantheonCommand.java (hunk 2 above)
Comment (abdelhamidbakhta) [reply pantheon/src/main/java/tech/pegasys/pantheon/cli/PantheonCommand.java]: Back to original place

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java [hunk 4]
```diff
@@ -845,6 +861,7 @@ public void setCurrentOperation(final Operation currentOperation) {
     private Address miningBeneficiary;
     private BlockHashLookup blockHashLookup;
     private Boolean isPersistingState = false;
+    private String reason = null;
```
[the remark this answers is not in the document]
Comment (abdelhamidbakhta) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java]: Done

[the remark this answers is not in the document]
Comment (mbaxter) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java]: +1 We are not 100% consistent on this, but `Optional`'s are generally preferred in this type of situation.

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java [hunk 5]
```diff
@@ -29,10 +29,7 @@
 import tech.pegasys.pantheon.util.uint.UInt256;
 import tech.pegasys.pantheon.util.uint.UInt256Value;
 
-import java.util.Deque;
-import java.util.EnumSet;
-import java.util.HashSet;
-import java.util.Set;
+import java.util.*;
```
Comment (mbaxter) [inline ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java]: We have a coding standard around not collapsing imports.

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (abdelhamidbakhta) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java]: Yes i agree

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java (hunk 5 above)
Comment (abdelhamidbakhta) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java]: This is my IDE, when the number of imports of the same package is higher than a threshold it replaces with *. i need to disable this

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/DebugTraceTransaction.java [hunk 6]
```diff
@@ -54,19 +56,31 @@ public JsonRpcResponse response(final JsonRpcRequest request) {
...
     final Optional<TransactionWithMetadata> transactionWithMetadata =
         blockchain.transactionByHash(hash);
-    final Hash blockHash = transactionWithMetadata.get().getBlockHash();
+    if (transactionWithMetadata.isPresent()) {
+      DebugTraceTransactionResult debugTraceTransactionResult =
+          debugTraceTransactionResult(hash, transactionWithMetadata.get(), transactionTraceParams);
+
+      return new JsonRpcSuccessResponse(request.getId(), debugTraceTransactionResult);
+    } else {
+      return new JsonRpcErrorResponse(request.getId(), JsonRpcError.TX_NOT_FOUND);
```
Comment (mbaxter) [inline ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/DebugTraceTransaction.java]: I don't think this is an error - you should just return a successful response with a `null` value.

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java (hunk 5 above)
Comment (mbaxter) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java]: yeah - i set mine to like 9999 or something :P

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/debug/TraceFrame.java [hunk 7]
```diff
@@ -34,6 +34,7 @@
   private final Optional<Bytes32[]> stack;
   private final Optional<Bytes32[]> memory;
   private final Optional<Map<UInt256, UInt256>> storage;
+  private final String reason;
```
Comment (mbaxter) [inline ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/debug/TraceFrame.java]: (recommended) How about `revertReason`?

File: ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/DebugTraceTransactionTest.java [hunk 8]
```diff
@@ -85,7 +85,8 @@ public void shouldTraceTheTransactionUsingTheTransactionTracer() {
             EnumSet.noneOf(ExceptionalHaltReason.class),
             Optional.empty(),
             Optional.empty(),
-            Optional.empty());
+            Optional.empty(),
+            "revert message");
```
Comment (mbaxter) [inline ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/DebugTraceTransactionTest.java]: Can you add a test for the case where no transaction is found?

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java (hunk 5 above)
Comment (abdelhamidbakhta) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java]: fixed

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/DebugTraceTransaction.java (hunk 6 above)
Comment (abdelhamidbakhta) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/DebugTraceTransaction.java]: Ok

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/debug/TraceFrame.java (hunk 7 above)
Comment (abdelhamidbakhta) [reply ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/debug/TraceFrame.java]: Ok

File: ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java (hunk 5 above)
Comment (shemnon) [inline ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/vm/MessageFrame.java]: This is triggering ErrorProne: https://jenkins.pegasys.tech/job/Pantheon/job/PR-1258/5/execution/node/38/log/

File: ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/DebugTraceTransactionTest.java (hunk 8 above)
Comment (abdelhamidbakhta) [reply ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/DebugTraceTransactionTest.java]: Done
